### ISSUE TITLE
Improper tilde in CDPATH causes segfault

### ISSUE THREAD

**Issue by fourjay** [neovim/neovim#2339]
Trigger
export CDPATH=~bogus
where ~bogus is not a valid path

Causes a segfault. Replicated with several versions of nvim and several platforms.

**Comment by splinterofchaos** [comment]
gdb output:
```
#0  0x00000000005fd0ea in kl_shift_Event (kl=0x0, d=0x7fffffffcf40) at /home/admin/src/neovim/src/nvim/os/event.c:29
#1  0x00000000005fd4cf in process_events_from (queue=0x0) at /home/admin/src/neovim/src/nvim/os/event.c:181
#2  0x00000000005fd3a1 in event_poll (ms=-1) at /home/admin/src/neovim/src/nvim/os/event.c:142
#3  0x00000000005faf34 in job_wait (job=0x8d7070, ms=-1) at /home/admin/src/neovim/src/nvim/os/job.c:246
#4  0x00000000005f5eab in shell (cmd=0x8d6ee0 "unset nonomatch; vimglob() { while [ $# -ge 1 ]; do echo \"$1\"; shift; done }; vimglob >/tmp/nvimPbuoX8/0 ~bogus", extra_args=0x0, input=0x0, len=0, output=0x0, 
    nread=0x7fffffffd260, silent=false, forward_output=false) at /home/admin/src/neovim/src/nvim/os/shell.c:245
#5  0x00000000005f5adc in os_call_shell (cmd=0x8d6ee0 "unset nonomatch; vimglob() { while [ $# -ge 1 ]; do echo \"$1\"; shift; done }; vimglob >/tmp/nvimPbuoX8/0 ~bogus", 
    opts=(kShellOptExpand | kShellOptSilent | kShellOptHideMess), extra_args=0x0) at /home/admin/src/neovim/src/nvim/os/shell.c:122
#6  0x0000000000522eeb in call_shell (cmd=0x8d6ee0 "unset nonomatch; vimglob() { while [ $# -ge 1 ]; do echo \"$1\"; shift; done }; vimglob >/tmp/nvimPbuoX8/0 ~bogus", 
    opts=(kShellOptExpand | kShellOptSilent | kShellOptHideMess), extra_shell_arg=0x0) at /home/admin/src/neovim/src/nvim/misc2.c:314
#7  0x000000000045c65f in mch_expand_wildcards (num_pat=1, pat=0x7fffffffd4a0, num_file=0x7fffffffd6ec, file=0x7fffffffd6f0, flags=43) at /home/admin/src/neovim/src/nvim/os_unix.c:423
#8  0x0000000000579264 in gen_expand_wildcards (num_pat=1, pat=0x7fffffffd4a0, num_file=0x7fffffffd6ec, file=0x7fffffffd6f0, flags=43) at /home/admin/src/neovim/src/nvim/path.c:1067
#9  0x000000000057ad8f in expand_wildcards (num_pat=1, pat=0x7fffffffd4a0, num_file=0x7fffffffd6ec, file=0x7fffffffd6f0, flags=43) at /home/admin/src/neovim/src/nvim/path.c:1885
#10 0x000000000057ad1d in expand_wildcards_eval (pat=0x7fffffffd4f0, num_file=0x7fffffffd6ec, file=0x7fffffffd6f0, flags=43) at /home/admin/src/neovim/src/nvim/path.c:1856
#11 0x000000000052ac07 in ExpandFromContext (xp=0x7fffffffd6c0, pat=0x8d5eb0 "~bogus", num_file=0x7fffffffd6ec, file=0x7fffffffd6f0, options=80) at /home/admin/src/neovim/src/nvim/ex_getln.c:3612
#12 0x0000000000529131 in ExpandOne (xp=0x7fffffffd6c0, str=0x8d5eb0 "~bogus", orig=0x0, options=80, mode=2) at /home/admin/src/neovim/src/nvim/ex_getln.c:2775
#13 0x00000000005fbe7f in expand_env_esc (srcp=0x8c9141 <NameBuff+1> "~bogus", dst=0x8d5eb0 "~bogus", dstlen=4095, esc=false, one=true, startstr=0x0) at /home/admin/src/neovim/src/nvim/os/env.c:302
#14 0x00000000005fbad4 in expand_env_save_opt (src=0x8c9141 <NameBuff+1> "~bogus", one=true) at /home/admin/src/neovim/src/nvim/os/env.c:183
#15 0x00000000005793ba in gen_expand_wildcards (num_pat=1, pat=0x7fffffffd850, num_file=0x7fffffffda9c, file=0x7fffffffdaa0, flags=43) at /home/admin/src/neovim/src/nvim/path.c:1105
#16 0x000000000057ad8f in expand_wildcards (num_pat=1, pat=0x7fffffffd850, num_file=0x7fffffffda9c, file=0x7fffffffdaa0, flags=43) at /home/admin/src/neovim/src/nvim/path.c:1885
#17 0x000000000057ad1d in expand_wildcards_eval (pat=0x7fffffffd8a0, num_file=0x7fffffffda9c, file=0x7fffffffdaa0, flags=43) at /home/admin/src/neovim/src/nvim/path.c:1856
#18 0x000000000052ac07 in ExpandFromContext (xp=0x7fffffffda70, pat=0x8c9141 <NameBuff+1> "~bogus", num_file=0x7fffffffda9c, file=0x7fffffffdaa0, options=80) at /home/admin/src/neovim/src/nvim/ex_getln.c:3612
#19 0x0000000000529131 in ExpandOne (xp=0x7fffffffda70, str=0x8c9141 <NameBuff+1> "~bogus", orig=0x0, options=80, mode=2) at /home/admin/src/neovim/src/nvim/ex_getln.c:2775
#20 0x00000000005fbe7f in expand_env_esc (srcp=0x8d2280 ",~bogus", dst=0x8c9141 <NameBuff+1> "~bogus", dstlen=4094, esc=false, one=false, startstr=0x0) at /home/admin/src/neovim/src/nvim/os/env.c:302
#21 0x000000000056b835 in option_expand (opt_idx=31, val=0x8d2280 ",~bogus") at /home/admin/src/neovim/src/nvim/option.c:3266
#22 0x0000000000568613 in set_init_1 () at /home/admin/src/neovim/src/nvim/option.c:1950
#23 0x00000000004ef912 in early_init () at /home/admin/src/neovim/src/nvim/main.c:178
#24 0x00000000004ef99e in main (argc=1, argv=0x7fffffffddd8) at /home/admin/src/neovim/src/nvim/main.c:202
```
This happens because `event_poll()` runs on this codepath before `event_init()`. This patch prevents the segfault for me (naive solution).
```patch
diff --git a/src/nvim/main.c b/src/nvim/main.c
index a03fd2e..86fa229 100644
--- a/src/nvim/main.c
+++ b/src/nvim/main.c
@@ -175,8 +175,6 @@ void early_init(void)
   // msg_outtrans_len_attr().
   // First find out the home directory, needed to expand "~" in options.
   init_homedir();               // find real value of $HOME
-  set_init_1();
-  TIME_MSG("inits 1");
 
   set_lang_var();               // set v:lang and v:ctype
 }
@@ -243,6 +241,9 @@ int main(int argc, char **argv)
   t_colors = 256;
   check_tty(&params);
 
+  set_init_1();
+  TIME_MSG("inits 1");
+
   /*
    * Set the default values for the options that use Rows and Columns.
    */
```
A quick scan of `event_init()` shows that most of the functionality does not seem to depend on the state of any other modules, so we can perhaps make an `event_early_init()` to run before `set_init_1()`, as a more permanent solution. Not going to work on this tonight, though.

**Comment by splinterofchaos** [comment]
I amended a commit to #2307 that fixes this. I decided to just add a check to `event_poll()` that `immediate_events != NULL` because there may be more bugs now or in the future caused by `_poll()` running before `_init()`, and changing the initialization order can have profound effects.

On the other hand, if we must worry about events being enqueued before `event_init()`, then `event_early_init()` might be more sane.
